kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt

```diff
@@ -307,30 +356,87 @@ internal open class CancellableContinuationImpl<in T>(
         dispatch(mode)
     }
 
-    // returns null when successfully dispatched resumed, CancelledContinuation if too late (was already cancelled)
-    private fun resumeImpl(proposedUpdate: Any?, resumeMode: Int): CancelledContinuation? {
+    private fun resumedState(
+        state: NotCompleted,
+        proposedUpdate: Any?,
+        resumeMode: Int,
+        onCancellation: ((cause: Throwable) -> Unit)?,
+        idempotent: Any?
+    ): Any? = when {
+        proposedUpdate is CompletedExceptionally -> {
+            assert { idempotent == null } // there are no idempotent exceptional resumes
+            assert { onCancellation == null } // only successful results can be cancelled
+            proposedUpdate
+        }
+        !resumeMode.isCancellableMode && idempotent == null -> proposedUpdate // cannot be cancelled in process, all is fine
+        onCancellation != null || state is CancelHandler || idempotent != null ->
+            // mark as CompletedContinuation if special cases are present (cancellation handlers or idempotent resume)
+            CompletedContinuation(proposedUpdate, state as? CancelHandler, onCancellation, idempotent)
```

Comment:
TBD: WARNING: A `CancellableContinuation` with result and a `CancelHandler` (that was installed via `invokeOnCancellation`) now becomes more expensive, since now `CompletedContinuation` wrapper is created in this case. It might affect performance of channels, which use `invokeOnCancellation` and might need a special workaround here, since we know that channels do not need "extended" `invokeOnCancellation` contract.

Reply:
It does not seem to produce a visible impact on benchmarks